src/DataReader/proto_reader.rs

```diff
@@ -72,202 +81,175 @@ pub fn proto_state_to_state(state: ProtoState, system: &TransitionSystemPtr) ->
     let federation: OwnedFederation =
         proto_federation_to_owned_federation(proto_federation, system);
 
-    let proto_location_tuple: ProtoLocationTuple = state.location_tuple.unwrap();
-    let location_tuple =
-        proto_location_tuple_to_location_tuple(&proto_location_tuple, system).unwrap();
+    let proto_location_tuple: LocationTree = state.location_tuple.unwrap();
+    let location_tuple = proto_location_tuple_to_location_tuple(proto_location_tuple, system);
+
+    // Ensure that the invariants are applied to the state
+    let federation = location_tuple.apply_invariants(federation);
 
     State::create(location_tuple, federation)
 }
 
 fn proto_location_tuple_to_location_tuple(
-    location_tuple: &ProtoLocationTuple,
+    location_tuple: LocationTree,
```

Comment:
Since, it is no longer a `LocationTuple`, I think it would be reasonable to rename `location_tuple` to `location_tree`
```suggestion
fn proto_location_tree_to_location_tree(
    location_tree: LocationTree,
```

Reply:
I have renamed everything to LocationTree now. Also variables and fields so we don't reference tuples anymore 🙂